solver/app/quote.go

```diff
@@ -1,63 +1,217 @@
 package app
 
 import (
+	"context"
+	"encoding/json"
 	"math/big"
+	"net/http"
 
 	"github.com/omni-network/omni/lib/errors"
+	"github.com/omni-network/omni/lib/ethclient/ethbackend"
+	"github.com/omni-network/omni/lib/log"
+
+	"github.com/ethereum/go-ethereum/common"
 )
 
+// NOTE: Quote request / response types mirror SolvertNet.OrderData, built
+// specifically for EVM -> EVM orders via SolverNetInbox / Outbox contracts,
+// with ERC7683 type hash matching SolverNetInbox.ORDERDATA_TYPEHASH.
+//
+// To support multiple order types with this api (e.g. EVM -> Solana, Solana -> EVM)
+// we'd need a more generic request / response format that discriminates on
+// order type hash.
+
+// QuoteRequest is the expected request body for the /api/v1/quote endpoint.
+type QuoteRequest struct {
+	SourceChainID      uint64         `json:"sourceChainId"`
+	DestinationChainID uint64         `json:"destChainId"`
+	FillDeadline       uint64         `json:"fillDeadline"`
+	Calls              []Call         `json:"calls"`
+	Expenses           []Expense      `json:"expenses"`
+	DepositToken       common.Address `json:"depositToken"`
+}
+
+// QuoteResponse is the response json for the /quote endpoint.
+type QuoteResponse struct {
+	Rejected          bool     `json:"rejected,omitempty"`
+	RejectReason      string   `json:"rejectReason,omitempty"`
+	RejectDescription string   `json:"rejectDescription,omitempty"`
+	Deposit           *Deposit `json:"deposit,omitempty"`
+	Error             *Error   `json:"error,omitempty"`
+}
+
+func (r QuoteResponse) StatusCode() int {
+	if r.Error != nil {
+		return r.Error.Code
+	}
+
+	return http.StatusOK
+}
+
+// Expense is a solver expense on the destination (matches bindings.SolverNetExpense).
+type Expense struct {
+	Spender common.Address `json:"spender"`
+	Token   common.Address `json:"token"`
+	Amount  *big.Int       `json:"amount"`
+}
+
+// Call is a call to be made on the destination (matches bindings.SolverNetCall).
+type Call struct {
+	Target   common.Address `json:"target"`
+	Selector [4]byte        `json:"selector"`
+	Value    *big.Int       `json:"value"`
+	Params   []byte         `json:"params"`
+}
+
+// Deposit is a user deposit on the source (matches bindings.SolverNetDeposit).
+type Deposit struct {
+	Token  common.Address `json:"token"`
+	Amount *big.Int       `json:"amount"`
+}
+
+// Error is a json response for http errors (e.g 4xx, 5xx), not used for rejections.
+type Error struct {
+	Code    int    `json:"code"`
+	Status  string `json:"status"`
+	Message string `json:"message"`
+}
+
+type quoteFunc func(context.Context, QuoteRequest) (Deposit, error)
+
+// newQuoter returns a quoteFunc that can be used to quote deposits for expenses.
+// It is the logic behind the /quote endpoint.
+func newQuoter(backends ethbackend.Backends, solverAddr common.Address) quoteFunc {
```

Comment:
Split logic from http wrangling. This simplifies testing which then doesn't need to do http stuff and can focus on logic.